src/wasm/wasm-type.cpp

```diff
@@ -204,82 +204,150 @@ template<> class hash<wasm::RecGroupStructure> {
   }
 };
 
+template<> class hash<wasm::TypeInfo> {
+public:
+  size_t operator()(const wasm::TypeInfo& info) const;
+};
+
+template<typename T> class hash<reference_wrapper<const T>> {
+public:
+  size_t operator()(const reference_wrapper<const T>& ref) const {
+    return hash<T>{}(ref.get());
+  }
+};
+
+template<typename T> class equal_to<reference_wrapper<const T>> {
+public:
+  bool operator()(const reference_wrapper<const T>& a,
+                  const reference_wrapper<const T>& b) const {
+    return equal_to<T>{}(a.get(), b.get());
+  }
+};
+
 } // namespace std
 
 namespace wasm {
 namespace {
 
+HeapTypeInfo* getHeapTypeInfo(HeapType ht) {
+  assert(!ht.isBasic());
+  return (HeapTypeInfo*)ht.getID();
+}
+
+HeapType asHeapType(std::unique_ptr<HeapTypeInfo>& info) {
+  return HeapType(uintptr_t(info.get()));
+}
+
+Type markTemp(Type type) {
+  if (!type.isBasic()) {
+    Type::getTypeInfo(type)->isTemp = true;
+  }
+  return type;
+}
+
+bool isTemp(Type type) {
+  return !type.isBasic() && Type::getTypeInfo(type)->isTemp;
+}
+
+bool isTemp(HeapType type) {
+  return !type.isBasic() && getHeapTypeInfo(type)->isTemp;
+}
+
 // Generic utility for traversing type graphs. The inserted roots must live as
 // long as the Walker because they are referenced by address. This base class
 // only has logic for traversing type graphs; figuring out when to stop
 // traversing the graph and doing useful work during the traversal is left to
-// subclasses.
+// subclasses, which should override `scanType` and/or `scanHeapType`. Edges
+// from reference types to the referenced heap types are not walked, so
+// subclasses should handle referenced heap types when their reference types are
+// visited.
 template<typename Self> struct TypeGraphWalkerBase {
-  void walkRoot(Type* type);
-  void walkRoot(HeapType* ht);
+  void walkRoot(Type* type) {
+    assert(taskList.empty());
+    taskList.push_back(Task::scan(type));
+    doWalk();
+  }
```

Comment:
As we don't walk heap types, it seems like this will only ever walk the type itself, or the tuple elements if it is a tuple. And the tuple elements are not themselves tuples. So can this function not just be a simple loop over `*type`?

Reply:
It could, but I think it's simpler to keep the method of traversal consistent between types and heap types.

Reply:
Fair enough, though it might be a tad faster the other way. But up to you.